File: problem-11/problem-11.go
package main

import (
	"os"
	"strconv"
	"strings"
)

// Size of 2D matrix
const SIZE = 20

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func load2DArray() [][]int {
	result := make([][]int, SIZE)

	// open file with sequence
	file, err := os.ReadFile("numbers.txt")
	check(err)

	// Split on newline
	lines := strings.Split(string(file), "\n")

	for i, line := range lines {
		// Create sub-array
		result[i] = make([]int, SIZE)

		digits := strings.Split(line, " ")

		for j, digit := range digits {
			result[i][j], err = strconv.Atoi(digit)
			check(err)
		}
	}

	return result
}

func main() {
	matrix := load2DArray()

	// We are calculating the 4 adjacent numbers, so TRAVERSE_SIZE reflects the bounds of this traversal
	const DISPLACE_SIZE = 3
	const TRAVERSE_SIZE = SIZE - DISPLACE_SIZE

	maxProduct := 0

	for row := range SIZE {
		for col := range SIZE {
			if col < TRAVERSE_SIZE {
				horizontalProduct := matrix[row][col] * matrix[row][col+1] * matrix[row][col+2] * matrix[row][col+3]
				if horizontalProduct > maxProduct {
					maxProduct = horizontalProduct
				}
			}

			if row < TRAVERSE_SIZE {
				verticalProduct := matrix[row][col] * matrix[row+1][col] * matrix[row+2][col] * matrix[row+3][col]
				if verticalProduct > maxProduct {
					maxProduct = verticalProduct
				}
			}

			// Diagonal with negative slope
			if col < TRAVERSE_SIZE && row < TRAVERSE_SIZE {
				diagonalProduct := matrix[row][col] * matrix[row+1][col+1] * matrix[row+2][col+2] * matrix[row+3][col+3]
				if diagonalProduct > maxProduct {
					maxProduct = diagonalProduct
				}
			}

			// Diagonal with positive slope
			if col >= DISPLACE_SIZE && row < TRAVERSE_SIZE {
				diagonalProduct := matrix[row][col] * matrix[row+1][col-1] * matrix[row+2][col-2] * matrix[row+3][col-3]
				if diagonalProduct > maxProduct {
					maxProduct = diagonalProduct
				}
			}
		}
	}

	println(maxProduct)
}

File: problem-11/c-cpp/problem-11.cpp
#include <iostream>
#include <fstream>

#include "../../utils/cpp/strings.cpp"

using namespace std;

int main() {
    // Open file
    ifstream inputFile("problem-11/numbers.txt");

    if (!inputFile.is_open()) {
        cerr << "Error: Could not open file." << endl;
        return 1;
    }

    vector<vector<int>> matrix;

    // Fill matrix with values
    string line;
    while (getline(inputFile, line)) {
        vector<string> numberStrs = StringSplit(line, " ");
        vector<int> numbers;
        for (auto numberStr : numberStrs) {
            numbers.push_back(stoi(numberStr));
        }

        matrix.push_back(numbers);
    }

    const int SIZE = 20;
    const int DISPLACE_SIZE = 3;
    const int TRAVERSE_SIZE = SIZE - DISPLACE_SIZE;

    int maxProduct = 0;

    for (int r = 0; r < SIZE; r++) {
        for (int c = 0; c < SIZE; c++) {
            if (c < TRAVERSE_SIZE) {
                const int horizontalProduct = matrix[r][c] * matrix[r][c+1] * matrix[r][c+2] * matrix[r][c+3];
                if (horizontalProduct > maxProduct) {
                    maxProduct = horizontalProduct;
                }
            }

            if (r < TRAVERSE_SIZE) {
                const int verticalProduct = matrix[r][c] * matrix[r+1][c] * matrix[r+2][c] * matrix[r+3][c];
                if (verticalProduct > maxProduct) {
                    maxProduct = verticalProduct;
                }
            }

            if (c < TRAVERSE_SIZE && r < TRAVERSE_SIZE) {
                const int diagonalProduct = matrix[r][c] * matrix[r+1][c+1] * matrix[r+2][c+2] * matrix[r+3][c+3];
                if (diagonalProduct > maxProduct) {
                    maxProduct = diagonalProduct;
                }
            }

            if (c >= DISPLACE_SIZE && r < TRAVERSE_SIZE) {
                const int diagonalProduct = matrix[r][c] * matrix[r+1][c-1] * matrix[r+2][c-2] * matrix[r+3][c-3];
                if (diagonalProduct > maxProduct) {
                    maxProduct = diagonalProduct;
                }
            }
        }
    }

    cout << maxProduct << endl;

    // Close file
    inputFile.close();

    return 0;
}